src/main.zig

```diff
@@ -2200,6 +2206,10 @@ pub fn cmdBuild(gpa: *Allocator, arena: *Allocator, args: []const []const u8) !v
                         override_global_cache_dir = args[i];
                         try child_argv.appendSlice(&[_][]const u8{ arg, args[i] });
                         continue;
+                    } else if (mem.eql(u8, arg, "--system-linker-hack")) {
+                        system_linker_hack = true;
+                        try child_argv.append(arg);
+                        continue;
```

Comment:
I don't like this being part of the `zig build` options. Under what conditions is it necessary for `zig build` to work? It sounds like you're saying not even "hello world" works on aarch64-macos without this.

Typically, we would do a workaround like this to unblock users from making progress on their projects while we work on solving the underlying issue - but since this system is not even close to being widely available yet, I think we should be in the business of either improving LLD or improving the self-hosted linker, rather than in providing workarounds for this experimental system. It's entirely possible that Apple reverses the decision and reverts the code-signing thing, or that LLD gets upstream support for this by LLVM 12. I'm already uncomfortable with the existence of --system-linker-hack for build artifacts, and I really hate to see it as a `zig build` option.

Now, that said, I want to recognize the following things:
 * You have put a lot of time and energy into understanding macos, mach-o, and the ecosystem
 * You have contributed significant, meaningful improvements to the self-hosted compiler on the related components
 * You have spent your own money to purchase a dev kit to test out the new Apple aarch64 thing
 * You do really high quality work, including collaborating with others

So I think you should take this code review as suggestions only, because I trust your judgement and leadership on this, and I think you should have some autonomy here to drive this user experience as you see fit.

Reply:
> but since this system is not even close to being widely available yet

FWIW Apple Silicon computers will be announced next Tuesday, so they will be widely available way before we cut another zig release.

Not even "hello world" works without this. And zig itself cannot even be compiled without this either.

Reply:
> I don't like this being part of the `zig build` options. Under what conditions is it necessary for `zig build` to work? It sounds like you're saying not even "hello world" works on aarch64-macos without this.
> 
> Typically, we would do a workaround like this to unblock users from making progress on their projects while we work on solving the underlying issue - but since this system is not even close to being widely available yet, I think we should be in the business of either improving LLD or improving the self-hosted linker, rather than in providing workarounds for this experimental system. It's entirely possible that Apple reverses the decision and reverts the code-signing thing, or that LLD gets upstream support for this by LLVM 12. I'm already uncomfortable with the existence of --system-linker-hack for build artifacts, and I really hate to see it as a `zig build` option.

I totally agree and I don't like it either. I'm gonna try your env var suggestion. I reckon this should work too and should be far less messy than cascading `--system-linker-hack` through the entire build tree.

> 
> Now, that said, I want to recognize the following things:
> 
> * You have put a lot of time and energy into understanding macos, mach-o, and the ecosystem
> * You have contributed significant, meaningful improvements to the self-hosted compiler on the related components
> * You have spent your own money to purchase a dev kit to test out the new Apple aarch64 thing
> * You do really high quality work, including collaborating with others
> 
> So I think you should take this code review as suggestions only, because I trust your judgement and leadership on this, and I think you should have some autonomy here to drive this user experience as you see fit.

Thanks, means the world to me!